File: OffOff_iOS/OffOff_iOS/Models/Post.swift
//
//  Post.swift
//  OffOff_iOS
//
//  Created by Lee Nam Jun on 2021/07/05.
//

struct PostList: Codable {
    var lastPostId: String
    var postList: [PostModel]
}

struct Author: Codable {
    var _id: String?
    var nickname: String
    var type: String?
    var profileImage: String?
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(_id, forKey: ._id)
        try container.encode(nickname, forKey: .nickname)
        try container.encode(type, forKey: .type)
        try container.encode(profileImage, forKey: .profileImage)
    }
}

struct PostModel: Codable {
    var _id: String?
    var boardType: String
    var author: Author
    var date: String
    var title: String
    var content: String
    var image: String?
    var likes: [String]
    var views: Int
    var reports: [String]
    var bookmarks: [String]
    var replyCount: Int
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(_id, forKey: ._id)
        try container.encode(author, forKey: .author)
        try container.encode(boardType, forKey: .boardType)
        try container.encode(content, forKey: .content)
        try container.encode(date, forKey: .date)
        try container.encode(image, forKey: .image)
        try container.encode(likes, forKey: .likes)
        try container.encode(reports, forKey: .reports)
        try container.encode(title, forKey: .title)
        try container.encode(views, forKey: .views)
        try container.encode(bookmarks, forKey: .bookmarks)
        try container.encode(replyCount, forKey: .replyCount)
    }
}

struct PostActivity: Codable {
    var boardType: String
    var postId: String
    var replyId: String
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        
        try container.encode(boardType, forKey: .boardType)
        try container.encode(postId, forKey: .postId)
        try container.encode(replyId, forKey: .replyId)
    }
}

File: OffOff_Android/app/src/main/java/com/yuuuzzzin/offoff_android/service/models/Post.kt
package com.yuuuzzzin.offoff_android.service.models

import com.google.gson.annotations.SerializedName
import java.io.Serializable

// 게시물 관련 객체 모델

/* 게시물 리스트 */
data class PostList(
    @SerializedName("lastPostId")
    val lastPostId: String,
    @SerializedName("postList")
    val postList: List<Post>
)

/* 게시물 */
data class Post(
    @SerializedName("_id")
    val id: String = "",
    @SerializedName("boardType")
    val boardType: String,
    @SerializedName("author")
    val author: Author,
    @SerializedName("title")
    val title: String,
    @SerializedName("content")
    val content: String,
    @SerializedName("image")
    val image: List<Image>? = emptyList(),
    @SerializedName("date")
    val date: String,
    @SerializedName("views")
    val views: Int? = 0,
    @SerializedName("replyCount")
    var replyCount: Int? = 0,
    @SerializedName("likes")
    val likes: List<String>? = emptyList(),
    @SerializedName("reports")
    val reports: List<Any>? = emptyList(),
    @SerializedName("bookmarks")
    val bookmarks: List<Any>? = emptyList(),
) : Serializable

/* 보내는 게시물 */
data class PostSend(
    @SerializedName("_id")
    val id: String? = null,
    @SerializedName("boardType")
    val boardType: String,
    @SerializedName("author")
    val author: String,
    @SerializedName("title")
    val title: String? = null,
    @SerializedName("content")
    val content: String? = null,
    @SerializedName("image")
    val image: List<Image>? = emptyList(),
)

/* 작성자 */
data class Author(
    @SerializedName("_id")
    val id: String,
    @SerializedName("nickname")
    val nickname: String? = null,
    @SerializedName("profileImage")
    val profileImage: List<Image>? = emptyList(),
    @SerializedName("type")
    val type: String? = null
) : Serializable

/* 이미지 */
data class Image(
    @SerializedName("key")
    val key: String?,
    @SerializedName("body")
    val body: String?
) : Serializable